wisl/lib/debugging/wislLifter.ml

```diff
@@ -1,152 +1,800 @@
+      | Update { d; _ }
+      | FunCall { d; _ }
+      | While { d; _ }
+      | Return { d } -> d
+
+    let update_partial_data { id; unifys; errors; _ } d =
+      d.ids |> ExtList.append id;
+      unifys |> List.iter (fun unify -> d.unifys |> ExtList.append unify);
+      errors |> List.iter (fun error -> d.errors |> ExtList.append error)
+
+    type partial_cmd_result =
+      | Finished of {
+          ids : rid list;
+          display : string;
+          unifys : unifys;
+          errors : string list;
+          cmd_kind : (branch_case, branch_data) cmd_kind;
+          submap : map submap;
+          fun_call_name : string option;
+        }
+      | StepAgain of (rid option * BranchCase.t option)
+      | NoPartial
+
+    let create annot tl_ast stmt exec_data =
+      let* origin_id = Annot.get_origin_id annot in
+      let* display = get_origin_node_str tl_ast (Some origin_id) in
+      let { cmd_report; _ } = exec_data in
+      let gil_cmd = CmdReport.(cmd_report.cmd) in
+      let failwith s =
+        DL.failwith
+          (fun () -> [ ("gil_cmd", cmd_to_yojson gil_cmd) ])
+          ("PartialCmds.create: " ^ s)
+      in
+      match exec_data.kind with
+      | Final -> None
+      | _ -> (
+          let d = make_partial_data display in
+          d |> update_partial_data exec_data;
+          if annot |> Annot.is_return then
+            Some (Return { d }, StepAgain (None, None))
+          else
+            let* stmt in
+            match stmt with
+            | WStmt.VarAssign (v, _) -> (
+                match gil_cmd with
+                | Cmd.Assignment (v', _) when v = v' -> None
+                | _ ->
+                    let partial = VarAssign { d; target_var = v } in
+                    Some (partial, StepAgain (None, None)))
+            | WStmt.Dispose _ -> (
+                match gil_cmd with
+                | Cmd.GuardedGoto _ ->
+                    let partial = Dispose { d } in
+                    Some
+                      ( partial,
+                        StepAgain (None, Some (BranchCase.GuardedGoto true)) )
+                | _ -> failwith "first cmd of Dispose wasn't GuardedGoto!")
+            | WStmt.Lookup (x, _) ->
+                let partial = Lookup { d; target_var = x } in
+                Some (partial, StepAgain (None, None))
+            | WStmt.Update _ -> (
+                match gil_cmd with
+                | Cmd.LAction (_, action, _)
+                  when WislLActions.(ac_from_str action = SetCell) ->
+                    failwith "first cmd of Update was SetCell!"
+                | Cmd.LAction (_, action, _)
+                  when WislLActions.(ac_from_str action = GetCell) ->
+                    let partial = Update { d; get_cmd_found = true } in
+                    Some (partial, StepAgain (None, None))
+                | _ ->
+                    let partial = Update { d; get_cmd_found = false } in
+                    Some (partial, StepAgain (None, None)))
```

Comment:
It feels like the abstraction could be made a bit nicer by capturing what's expected next instead of matching or something, I don't know.
What I do know is that I don't care too much because this is not Gillian, and that's what beautiful about it! Bring your own abstraction to the lifter.

Reply:
I'm very happy with "just works" for the time being :P 
I think the time to think harder on this is when GIL is abstracted away and we can no longer match on it.

Reply:
When Gillian is abstracted away, there is no more lifting going on, so that will be much easier actually :p